_posts/2024-08-19-given-priority-change-3.7.md

```diff
@@ -0,0 +1,232 @@
+---
+layout: blog-detail
+post-type: blog
+by: Oliver Bračevac, EPFL
+title: "Upcoming Changes to Givens in Scala 3.7"
+---
+
+## New Prioritization of Givens in Scala 3.7
+
+Scala 3.7 introduces changes to how `givens` are resolved, which can
+affect program behavior when multiple givens are present.
+
+For example, consider a library that provides a default
+`given` for a component:
+```scala
+// library code
+class LibComponent:
+    def msg = "library-defined"
+
+// default provided by library
+given libComponent: LibComponent = LibComponent()
+
+def printComponent(using c:LibComponent) = println(c.msg)
+```
+
+Up until Scala 3.6, clients of the library could override
+`libComponent` with a user-defined one through subtyping
+
+```scala
+// client code
+class UserComponent extends LibComponent:
+    override def msg = "user-defined"
+
+given userComponent: UserComponent = UserComponent()
+
+@main def run = printComponent
+```
+
+Let's run the example:
+
+```scala
+run // Scala <= 3.6: prints "user-defined"
+    // Scala 3.7: prints "library-defined"
+```
+
+What happened? In Scala 3.6 and earlier, the compiler prioritizes the
+`give`n with the _most specific_ compatible subtype
+(`userComponent`). However, in Scala 3.7, it selects the value with the
+_most general_ subtype instead (`libComponent`).
+
+
+### Motivation: Better Handling of Inheritance Triangles & Typeclasses
```

Comment:
In most of [Scala documentation](https://docs.scala-lang.org/scala3/book/ca-type-classes.html) we're using `Type Class` in opposition to `Typeclass` that [seems to be used by Haskell community](https://serokell.io/blog/haskell-typeclasses)
```suggestion
### Motivation: Better Handling of Inheritance Triangles & Type classes
```
There's more of typeclass usages in the text

Reply:
`Type class` is also the correct terminology here as in [typeclassopedia](https://wiki.haskell.org/Typeclassopedia):


> The original spelling of “type class” is with two words, as evidenced by, for example, the [Haskell 2010 Language Report](http://www.haskell.org/onlinereport/haskell2010/), early papers on type classes like [Type classes in Haskell](https://dl.acm.org/doi/pdf/10.1145/227699.227700) and [Type classes: an exploration of the design space](https://citeseer.ist.psu.edu/viewdoc/download?doi=10.1.1.1085.8703&rep=rep1&type=pdf), and [Hudak et al.’s history of Haskell](http://citeseer.ist.psu.edu/viewdoc/summary?doi=10.1.1.168.4008). However, as often happens with two-word phrases that see a lot of use, it has started to show up as one word (“typeclass”) or, rarely, hyphenated (“type-class”).

Reply:
fwiw (and at the risk of igniting a full-on holy war 😬), in Scala contexts I think "typeclass" is preferable because typeclasses aren't classes. Haskell doesn't have classes so there is no confusion in that context but in ours there is

Reply:
If [my boss](https://docs.scala-lang.net/sips/sips/typeclasses-syntax.html) uses ‘typeclass’, who am I to split words? 😄

Reply:
I don't feel too strongly about it, though I tend to agree with @SethTisue here. But if there's a majority that wants to have it changed to "Type Classes", I'll bend the knee.